contributors/design-proposals/runas-groupid.md

```diff
@@ -0,0 +1,191 @@
+     // +optional
+     RunAsGroup *int64
+
+    .....
+ }
+```
+
+#### PodSecurityPolicy
+
+PodSecurityPolicy defines strategies or conditions that a pod must run with in order to be accepted
+into the system. Two of the relevant strategies are RunAsUser and SupplementalGroups. We introduce 
+a new strategy called RunAsGroup which will support the following options:
+- MustRunAs
+- MustRunAsNonRoot
+- RunAsAny
+
+```
+// PodSecurityPolicySpec defines the policy enforced.
+ type PodSecurityPolicySpec struct {
+     //Other fields not shown for brevity
+    ..... 
+  // RunAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
+  RunAsUser RunAsUserStrategyOptions
+  // SupplementalGroups is the strategy that will dictate what supplemental groups are used by the SecurityContext.
+  SupplementalGroups SupplementalGroupsStrategyOptions
+
+
+  // RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+  RunAsGroup RunAsGroupStrategyOptions
+   .....
+}
+
+// RunAsGroupStrategyOptions defines the strategy type and any options used to create the strategy.
+ type RunAsUserStrategyOptions struct {
+     // Rule is the strategy that will dictate the allowable RunAsGroup values that may be set.
+     Rule RunAsGroupStrategy
+     // Ranges are the allowed ranges of gids that may be used.
+     // +optional
+     Ranges []GroupIDRange
+ }
+
+// RunAsGroupStrategy denotes strategy types for generating RunAsGroup values for a
+ // SecurityContext.
+ type RunAsGroupStrategy string
+ 
+ const (
+     // container must run as a particular gid.
+     RunAsGroupStrategyMustRunAs RunAsGroupStrategy = "MustRunAs"
+     // container must run as a non-root gid
+     RunAsGroupStrategyMustRunAsNonRoot RunAsGroupStrategy = "MustRunAsNonRoot"
+     // container may make requests for any gid.
+     RunAsGroupStrategyRunAsAny RunAsGroupStrategy = "RunAsAny"
+ )
+```
+
+## Behavior
+
+Following points should be noted:
+
+- `FSGroup` and `SupplementalGroups` will continue to have their old meanings and would be untouched.  
+- The `RunAsGroup` In the SecurityContext will override the `RunAsGroup` in the PodSecurityContext.
+- If no `RunAsGroup` is provided in the PodSecurityContext and SecurityContext, the Group provided 
+  in the Docker image will be used.
+- If no `RunAsGroup` is provided in the PodSecurityContext and SecurityContext, and none in the image,
+  the container will run with primary Group as root(0).
+
+## Summary of Changes needed
+
+At a high level, the changes classify into:
+1. API
+2. Validation
+3. CRI
+4. Runtime for Docker and rkt
+5. Swagger
+6. DockerShim
+7. Admission
+8. Registry
+
+- plugin/pkg/admission/security/podsecuritypolicy
+- plugin/pkg/admission/securitycontext
+- pkg/securitycontext/util.go
+- pkg/security/podsecuritypolicy/selinux
+- pkg/security/podsecuritypolicy/user
+- pkg/security/podsecuritypolicy/group
+- pkg/registry/extensions/podsecuritypolicy/storage
```

Comment:
I'm surprised this package has to change - are you sure?

Reply:
there are some tests which verify storage of PodSecurityPolicy at REST layer here https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/extensions/podsecuritypolicy/storage/storage_test.go this will change if we need to validate RunAsGroup field as well